src/dataframerow/utils.jl

```diff
@@ -140,21 +150,29 @@ function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
     return ngroups, rhashes, gslots, false
 end
 
-nlevels(x::PooledArray) = length(x.pool)
-nlevels(x) = length(levels(x))
-
-function row_group_slots(cols::NTuple{N,<:Union{CategoricalVector,PooledVector}},
+# Optimized method for arrays for which DataAPI.refpool is defined and returns an AbstractVector
+function row_group_slots(cols::NTuple{N,<:AbstractVector},
+                         refpools::NTuple{N,<:AbstractVector},
                          hash::Val{false},
                          groups::Union{Vector{Int}, Nothing} = nothing,
-                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool} where N
+                         skipmissing::Bool = false,
+                         sort::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool} where N
     # Computing neither hashes nor groups isn't very useful,
     # and this method needs to allocate a groups vector anyway
     @assert groups !== nothing && all(col -> length(col) == length(groups), cols)
 
+    refpools = map(DataAPI.refpool, cols)
+    foreach(refpool -> @assert(allunique(refpool)), refpools)
```

Comment:
This should probably be turned into a check, with a fallback to the standard method if it fails, but I wanted to discuss that are the expectations for a `refpool`. Should we require it to contain unique values or not? @quinnj @piever

Reply:
Hmmmm....in the arrow format, duplicate values are specifically allowed in the "dict encoding" pool, though it's recommended to "normalize" them to avoid duplicates, so I'm not sure; probably a fine requirement in practice, but could potentially cause an issue for weird arrow files.

Reply:
OK, so I'll turn this into a simple check with a fallback to the generic method.

Reply:
Actually, in some extreme cases it can take a lot of time. For example, the following takes 65s currently, but only 27s if I get rid of the check. So we probably should add an API to check whether the pool may contain duplicates. For `PooledArray` and `CategoricalArray` it would always be false, for others it could either always be true or vary at runtime.
```julia
using DataFrames, PooledArrays
N = 1_000_000_000;
k = N ÷ 5
df = DataFrame(x=PooledArray(rand(1:k, N)), y=rand(N));
@time gd = groupby(df, :x);
```

Reply:
`Arrow.DictEncoded` allows duplicates and would need a runtime check

Reply:
Yes that's why I mentioned that possibility.

Reply:
I wanted to use refpool for a different project, but the fact that refpool could have non-unique element made the code too complicated to be worth it. I agree it would be nice to have an API to check pools are unique.

Reply:
See also https://github.com/JuliaData/DataFrames.jl/issues/2558.